File: modules/Scripting/annotations/RmlUi/Event.lua
---@meta

---@enum ui.EventPhase
ui.EventPhase = {
  None = 0, Capture = 1, Target = 2, Bubble = 4,
}

---@class ui.Event
---@overload fun(): ui.Event
---@overload fun(target: ui.Element, id: ui.EventId, type: string, parameters: ui.Dictionary, interruptible: boolean): ui.Event
ui.Event = {}

---Get the current propagation phase.
---@return ui.EventPhase
function ui.Event:getPhase() end

---Set the current propagation phase.
---@param phase ui.EventPhase
function ui.Event:setPhase(phase) end

---Set the current element in the propagation.
---@param element ui.Element
function ui.Event:setCurrentElement(element) end

---Get the current element in the propagation.
---@return ui.Element
function ui.Event:getCurrentElement() end

---Get the target element of this event.
---@return ui.Element
function ui.Event:getTargetElement() end

---Get the event type.
---@return string
function ui.Event:getType() end

---Get the event id.
---@return ui.EventId
function ui.Event:getId() end

---Stops propagation of the event if it is interruptible, but finish all listeners on the current element.
function ui.Event:stopPropagation() end

---Stops propagation of the event if it is interruptible, including to any other listeners on the current element.
function ui.Event:stopImmediatePropagation() end

---Returns true if the event can be interrupted, that is, stopped from propagating.
---@return boolean
function ui.Event:isInterruptible() end

---Returns true if the event is still propagating.
---@return boolean
function ui.Event:isPropagating() end

---Returns true if the event is still immediate propagating.
---@return boolean
function ui.Event:isImmediatePropagating() end

---Returns the value of one of the event's parameters.
---@param key string # The name of the desired parameter.
---@return ui.Variant # The value of the requested parameter.
function ui.Event:getParameter(key) end

---Return the unprojected mouse screen position.<br/>
---NOTE: Only specified for events with 'mouse_x' and 'mouse_y' parameters.
---@return vec2
function ui.Event:getUnprojectedMouseScreenPos() end

File: modules/Scripting/modules/RmlUi/src/Event.cpp
#include "Event.hpp"
#include "RmlUi/Core/Element.h"
#include "Rml2glm.hpp"

#include "Sol2HelperMacros.hpp"

using namespace Rml;

void registerEvent(sol::table &lua) {
  // clang-format off
#define MAKE_PAIR(Value) _MAKE_PAIR(EventPhase, Value)
  DEFINE_ENUM(EventPhase, {
    MAKE_PAIR(None),
    MAKE_PAIR(Capture),
    MAKE_PAIR(Target),
    MAKE_PAIR(Bubble),
  });
#undef MAKE_PAIR

  DEFINE_USERTYPE(Event,
    sol::call_constructor,
    sol::constructors<
      Event(),
      Event(Element *, EventId, const String &, const Dictionary &, bool)
    >(),

    sol::meta_function::equal_to, sol::overload(
      sol::resolve<bool(const String &) const>(&Event::operator==),
      sol::resolve<bool(EventId) const>(&Event::operator==)
    ), 

    "getPhase", &Event::GetPhase,
    "setPhase", &Event::SetPhase,

    "setCurrentElement", &Event::SetCurrentElement,
    "getCurrentElement", &Event::GetCurrentElement,
    "getTargetElement", &Event::GetTargetElement,
    
    "getType", &Event::GetType,
    "getId", &Event::GetId,
    
    "stopPropagation", &Event::StopPropagation,
    "stopImmediatePropagation", &Event::StopImmediatePropagation,
    
    "isInterruptible", &Event::IsInterruptible,
    "isPropagating", &Event::IsPropagating,
    "isImmediatePropagating", &Event::IsImmediatePropagating,

    "getParameter", [](const Event &self, const String &name) {
      const auto &parameters = self.GetParameters();
      const auto it = parameters.find(name);
      return it != parameters.cend() ? it->second : Variant{};
    },

    "getUnprojectedMouseScreenPos", [](const Event &self) {
      return to_glm(self.GetUnprojectedMouseScreenPos());
    },

    BIND_TOSTRING(Event)
  );
  // clang-format on
}
